### PR TITLE
Separate parsing and core css functionality

### PR DESCRIPTION
**Pull request by WorldSEnder**
As mentioned in #8 this introduces two crates (macro part not implemented), one for the representation of css (`stylist-core`) and the other for parsing and the `yield_style` in (`stylist` itself).

### CODE REVIEW

File: packages/stylist-core/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    "HtmlHeadElement",
+    "HtmlStyleElement",
+] }
+
+[target.'cfg(target_arch = "wasm32")'.dev-dependencies]
+wasm-bindgen-test = "0.3.24"
+
+[features]
+default = []
+yew_integration = ["yew"]
```
Comment (WorldSEnder) [inline packages/stylist-core/Cargo.toml:49]: The feature name has been changed back to `yew_integration`, as having a feature and optional dependency of the same name is [currently not possible](https://github.com/rust-lang/cargo/issues/5565). Renaming the dependency internally to something like `yew_dep = { version = ..., package = "yew" }` messes with some `yew` macros, so  I didn't bother going that route.

File: packages/stylist-core/src/ast.rs [hunk 2]
```diff
@@ -49,18 +57,10 @@ impl ToCss for Scopes {
...
-
-#[derive(Debug, Clone, PartialEq)]
-pub(crate) struct Scope {
-    pub(crate) condition: Option<String>,
-    pub(crate) stylesets: Vec<ScopeContent>,
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Scope {
+    pub condition: Option<String>,
+    pub stylesets: Vec<ScopeContent>,
 }
```
Comment (WorldSEnder) [inline packages/stylist-core/src/ast.rs:64]: I'm still deliberating with myself if the fields should be public, or if we'd just want to expose public constructors (e.g. one with condition and one without) and mutable access for the stylesets part.

File: packages/stylist-core/src/registry.rs [hunk 3]
```diff
@@ -68,25 +61,23 @@ mod tests {
...
         init();
 
-        let style_a = Style::create("element-a", r#"color: red;"#)?;
-        let style_b = Style::create("element-b", r#"color: red;"#)?;
+        let style_a = Style::create_from_scopes("element-a", sample_scopes());
+        let style_b = Style::create_from_scopes("element-b", sample_scopes());
 
-        assert_ne!(style_a.get_style_str(), style_b.get_style_str());
-        Ok(())
+        assert_eq!(style_a.get_style_str(), style_b.get_style_str());
```
Comment (WorldSEnder) [inline packages/stylist-core/src/registry.rs:73]: Changed: equality in the StyleRegistry does *not* respect the classname. As this is a hint only, anyway, and an unpredictable rng string is appended; I don't see the point of caching by name.

One worry would be that users take the generated classname and insert additional css with some other method that then also effects deduplicated styles with the same scopes, but imo this should be pointed out in the documentation as bad practice and not be done. Rather one'd create an additional classname and add additional styles by combining multiple classes via the element's `class = "element-a element-b"` attribute.

File: packages/stylist-core/src/style.rs [hunk 4]
```diff
@@ -91,84 +86,84 @@ pub struct Style {
...
+            inner: Arc::new(StyleContent {
+                class_name: format!("{}-{}", class_prefix, get_rand_str()),
+                ast: key.0.clone(),
+                style_str: OnceCell::new(),
+                key: Arc::new(key),
+            }),
+        };
+
+        #[cfg(target_arch = "wasm32")]
+        new_style.inner.mount().expect("Failed to mount");
```
Comment (WorldSEnder) [inline packages/stylist-core/src/style.rs:112]: I'm not sure auto-mounting on creation is a good idea. For example, this would trigger when the application is started and loading, but the document is not ready yet? I'd rather let the user decide when to have the style mount/mounted/unmount.

File: packages/stylist-core/src/style.rs [hunk 5]
```diff
@@ -210,47 +206,10 @@ impl Style {
...
-
-@media screen and (max-width: 600px) {{
-.{style_name} {{
-color: yellow;
-}}
-}}
-"#,
-                style_name = style.get_class_name()
-            )
-        )
```
Comment (WorldSEnder) [inline packages/stylist-core/src/style.rs:254]: TODO: add these tests back in `stylist` instead of `stylist-core`.

**Comment by WorldSEnder** [comment]
@futursolo alright, can you have a look over it and tell me if the design is okay to you so far, if there is anything missing that you want to see, and what should be documented better?

File: packages/stylist-core/Cargo.toml (hunk 1 above)
Comment (futursolo) [reply packages/stylist-core/Cargo.toml:49]: If my understanding is correct, it's possible as long as you are not relying on additional optional dependencies.
You can try one of the examples in the examples folder on the master branch.

> Optional dependencies implicitly define a feature of the same name as the dependency. This means that the same cfg(feature = "gif") syntax can be used in the code, and the dependency can be enabled just like a feature such as --features gif (see Command-line feature options below).

https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies

File: packages/stylist-core/src/registry.rs (hunk 3 above)
Comment (futursolo) [reply packages/stylist-core/src/registry.rs:74]: It used to be the class name of the Style, but I have changed it to be the prefix.

The first item in the tuple is the prefix ("element-a" or "stylist" by default), not the generated class name with entropy.
Yes, it’s purely visual, but somewhat important for debugging.

The purpose of the test is to test that the class prefix provided to Style::create is respected.
(Although, maybe testing the class name here is better.)

```rust
    // assert_ne!(".element-a-lCzxVWIt", ".element-b-WjX1zwQ1”);
    assert_ne!(element_a.get_class_name(), element_b.get_class_name());
```

If StyleKey is changed to save the parsed AST only, it will only use the prefix of the first generated stylesheet for all following stylesheets that has the same AST (both Styles will have a prefix of "element-a").

File: packages/stylist-core/src/style.rs (hunk 4 above)
Comment (futursolo) [reply packages/stylist-core/src/style.rs:109]: I actually thought about this when I was rewriting the struct. But I figured that it's ok to do that for the moment as most of the time this is the desired behaviour and I want to maintain some API compatibility with `css-in-rust`. And wasm files are usually loaded with fetch by a javascript wrapper in `<body />` and by that time document head is probably ready.

With that being said though, I did saw #11. I will address after this PR is merged.

File: packages/stylist-core/src/style.rs (hunk 5 above)
Comment (futursolo) [reply packages/stylist-core/src/style.rs:254]: I am not sure if Style also needs to be moved to `stylist_core`.

Does the procedural macro also need to use Style at compile time?

If my understanding is correct, it's the generated code inside `quote!` that will use Style to construct from AST.

File: packages/stylist-core/Cargo.toml (hunk 1 above)
Comment (WorldSEnder) [reply packages/stylist-core/Cargo.toml:49]: Yeah, this is complicated by the fact that `stylist` itself does not (directly) depend on yew, but `stylist-core` does. So the feature flag in `stylist` should trigger the one in `stylist-core`. But alas, since `stylist` has doc-tests that depend on `yew`, it has a `dev-dependency` on `yew`, which is additionally not optional!

File: packages/stylist-core/src/registry.rs (hunk 3 above)
Comment (WorldSEnder) [reply packages/stylist-core/src/registry.rs:74]: If it's for debugging, I wonder if it could be better to maintain a list of original prefixes in debug mode and add them to a `data-` attribute of the generated style element? Just thought mounting multiple components for the same ast would be more wasteful.

File: packages/stylist-core/Cargo.toml (hunk 1 above)
Comment (futursolo) [reply packages/stylist-core/Cargo.toml:49]: image

I don't think that you need to list it as a dev-dependency. But if you want to change it to `yew-integration` it's fine with me as well.

Comment (WorldSEnder) [reply packages/stylist-core/Cargo.toml:49]: On master it works fine! With the proposed changes it would not. Can you make sure you have checked out this pull request when testing?

```
git fetch origin refs/pull/10/head
git checkout FETCH_HEAD
```

Here it shouldn't work when you remove yew from the dev-dependencies (since it also isn't in dependencies)

File: packages/stylist-core/src/style.rs (hunk 4 above)
Comment (WorldSEnder) [reply packages/stylist-core/src/style.rs:109]: Fine with me!

File: packages/stylist-core/src/style.rs (hunk 5 above)
Comment (WorldSEnder) [reply packages/stylist-core/src/style.rs:254]: Good point, maybe `Style` and the registry should live in `stylist` instead of `stylist-core`, that would also solve an issue I had with documentation not showing the `impl From<Styles> for Classes` - which does show for `pub type Style = stylist_core::Style` but not if one uses `pub use stylist_core::Style`. Additionally could get rid of `StyleExt`.

In that case `style-core` would contain only the ast representation, the macro generates a `Sheet`, but not a `Style` and the user calls the correct `Style::new`/`Style::create`.

Comment (futursolo) [reply packages/stylist-core/src/style.rs]: I think you can generate code that references to the main crate (Style) as long as you are not using them at the compile time.

As the macro and core will be reexport to stylist, and generated code should also only reference them via the re-exported path in the main crate so they don't have to list stylist-core as a dependency.

Example: https://github.com/yewstack/yew/blob/master/packages/yew-macro/src/classes/mod.rs#L30

Please correct me if I am wrong.

File: packages/stylist-core/Cargo.toml (hunk 1 above)
Comment (futursolo) [reply packages/stylist-core/Cargo.toml]: I see your point. Sorry about that.

File: packages/stylist/src/style_ext.rs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    /// # Examples
+    ///
+    /// ```rust
+    /// use stylist::{Style, StyleExt};
+    ///
+    /// let style = Style::create("my-component", "background-color: red;")?;
+    /// # use stylist::TryParseCss;
+    /// # Ok::<(), <&str as TryParseCss>::Error>(())
+    /// ```
+    fn create<N: Borrow<str>, Css: TryParseCss>(classname: N, css: Css)
```
Comment (futursolo) [inline packages/stylist/src/style_ext.rs]: What's the advantage of taking a custom trait over `AsRef<str>` for style sheet?

If it's a substitute for the `FromStr` trait on the original AST, maybe we can either move parser to stylist-core or just get away with it by using `Parser::parse` directly?

maybe prefix should be AsRef<str> as well?

File: packages/stylist-core/src/registry.rs (hunk 3 above)
Comment (futursolo) [reply packages/stylist-core/src/registry.rs]: I cannot imagine the performance impact will be too much (Once cached). And there's no impact if you don't use it.

This PR is getting very large. Can we keep the current behaviour and address this in another round (with #11, as that also requires changes to mounting) if it's of your concern?

**Review by futursolo — COMMENTED** [review]
@WorldSEnder

Thank you for your PR. I have written some thoughts on it.

File: packages/stylist/src/style_ext.rs (hunk 6 above)
Comment (WorldSEnder) [reply packages/stylist/src/style_ext.rs]: The point was to have different types of errors that are influenced by the parsing behavior. At first, I had one method and two impls, one for `&str`, another for `Style` that was Infallible and just cloned. I think in the course of the implementation I realized that most of the utility of Infallible isn't there yet, e.g. you can't easily with smth like [`Result::into_ok`](https://doc.rust-lang.org/std/result/enum.Result.html#method.into_ok) in stable rust, so I got rid of the second impl in favor of adding back the `new` constructors.

Tl;dr: Going back to `AsRef<str>` is fine by me.

Comment (WorldSEnder) [reply packages/stylist/src/style_ext.rs]: I would vote for having two different error types for parsing and mounting though

File: packages/stylist-core/src/style.rs [hunk 7]
```diff
@@ -91,84 +83,83 @@ pub struct Style {
...
     /// ```
-    /// use stylist::Style;
-    ///
-    /// let style = Style::new("color:red;").expect("Failed to create style.");
+    /// use stylist_core::Style;
     ///
-    /// // Example Output: stylist-uSu9NZZu
-    /// println!("{}", style.get_class_name());
+    /// let scopes = Default::default();
+    /// let style = Style::create_from_sheet("my-component", scopes);
```
[the remark this answers is not in the document]
Comment (WorldSEnder) [reply packages/stylist-core/src/style.rs]: The main purpose of these methods is to advertise that you can create a `Style` without having to unwrap a `Result`, if you don't want to go through the (fallible) parser. I will add a comment to the docs about that.

File: packages/stylist-core/src/registry.rs (hunk 3 above)
Comment (WorldSEnder) [reply packages/stylist-core/src/registry.rs]: Replaced with `assert_ne!(element_a.get_class_name(), element_b.get_class_name());` and added the string back to the StyleKey. Doing a copy here instead of storing a `Cow<...>`, compared to the cost of hashing and comparing the ast, that should be negligible.

**Comment by WorldSEnder** [comment]
I want to add some bits of documentation to `style-core` tomorrow now that the ast is semi-public, then mark it as ready for review.

File: packages/stylist-core/src/style.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (WorldSEnder) [reply packages/stylist-core/src/style.rs]: I also should mention that the macro will probably return a `Sheet`, not a style directly, so then you'd use these methods.

**Review by futursolo — APPROVED** [review]
